README.md
# Veronica

[![Build Status](https://travis-ci.com/bermannoah/veronica.svg?branch=master)](https://travis-ci.com/github/bermannoah/veronica)

A gem to make it easier for sardonic young detectives to keep secrets. 

Basically just a wrapper around existing hashing/encrypting methods.

Not really actively maintained, use at your own risk. May spontaneously accuse trusted community figures of murder, or simply not work.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'veronica'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install veronica

## Usage
Once you've included it in a project, you can use it like so:

```
Veronica.generate_sha2(digest_bits: 256, string: "I've got a secret, a good one.")`
=> "61cb3c4118402462e3055033176a2222d1a7b8e88db775bbdccd5b05b3cc7f0e"
```

You can also pass in an encoding_format:
```
Veronica.generate_sha2(digest_bits: 256, string: "Don't forget about me, Veronica.", encoding_format: "base64digest")
=> "wyWcnqQlnMtEdgR23nOeGoDm3Z5wetAj5FHoIqXj2d4="
```

Valid digest_bits settings for SHA2 are: `256, 384, 512`

Valid encoding_format settings for SHA2 are "hexdigest", "base64digest", and just "digest". 

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. Feel free to submit issues as well as PRs. 

## Contributing

Bug reports and pull requests are welcome on [GitHub](https://github.com/bermannoah/veronica/issues).
Please respect the [Code of Conduct](https://github.com/bermannoah/veronica/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.1] - 2021-06-15
### Added
- Updated gemfile for tests, bumped bundler version for builds

## [0.2.0] - 2021-06-09
### Security
- Patch insecure versions of several dependencies (veronica itself not affected)

## [0.1.0] - 2017-12-07
### Added
- Initial release


spec/spec_helper.rb
require 'simplecov'
SimpleCov.start do
  add_filter "/spec/"
end

require 'pry'
require 'zxcvbn'
require 'faker'
require_relative '../lib/veronica'

RSpec.configure do |config|
  config.order = 'random'
  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

end


spec/veronica_spec.rb
require 'spec_helper'

RSpec.describe Veronica do  
  it "has a version number" do
    expect(Veronica::VERSION).not_to be nil
  end
  
  it "is a module" do
    expect(Veronica).to be_a(Module)
  end
  
  it "raises if passed an invalid bit length for sha2" do
    expect {
      Veronica.generate_sha2(digest_bits: 13, string: "oh no")
    }.to raise_error(/You can only use 256, 384, or 512 for bit length/)
  end

  it "raises if passed an invalid encoding format for sha2" do
    expect {
      Veronica.generate_sha2(digest_bits: 256, string: "oh no", encoding_format: "pretty flowers")
    }.to raise_error(/You can only use hexdigest, base64digest or just digest for an encoding format/)
  end
  
  it "hashes a string with sha2 256" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 256, string: string)
    analysis = Zxcvbn.test(hash, [string])

    expect(hash).to_not eq(string)
    expect(analysis.crack_time_display).to eq("centuries")
  end

  it "hashes a string with sha2 384" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 384, string: string)
    analysis = Zxcvbn.test(hash, [string])

    expect(hash).to_not eq(string)
    expect(analysis.crack_time_display).to eq("centuries")
  end

  it "hashes a string with sha2 512" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 512, string: string)
    analysis = Zxcvbn.test(hash, [string])
    
    expect(hash).to_not eq(string)
    expect(analysis.crack_time_display).to eq("centuries")
  end

  it "hashes a string with sha2 base64digest" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 512, string: string, encoding_format: "base64digest")
    analysis = Zxcvbn.test(hash, [string])

    expect(hash).to_not eq(string)
    expect(analysis.crack_time_display).to eq("centuries")
  end

  it "hashes a string with sha2 digest" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 512, string: string, encoding_format: "digest")
    analysis = Zxcvbn.test(hash, [string])
    
    expect(hash).to_not eq(string)
    expect(analysis.crack_time_display).to eq("centuries")
  end

  it "hashes a string the same way twice" do
    string = Faker::Books::CultureSeries.culture_ship
    hash = Veronica.generate_sha2(digest_bits: 512, string: string)
    
    expect(hash).to_not eq(string)
    
    hash_again = Veronica.generate_sha2(digest_bits: 512, string: string)
    expect(hash).to eq(hash_again)
  end
end

lib/veronica.rb
require "veronica/version"
require 'digest'

module Veronica
  VALID_SHA2_BIT_LENGTHS = [256, 384, 512]
  VALID_SHA2_ENCODING_FORMATS = ["hexdigest", "base64digest", "digest"]

  # We use hexdigest by default for familiarities sake
  def self.basic_sha2_string(digest_bits: , string: , encoding_format: "hexdigest")
    unless VALID_SHA2_BIT_LENGTHS.include?(digest_bits)
      raise ArgumentError, "You can only use 256, 384, or 512 for bit length."
    end

    unless VALID_SHA2_ENCODING_FORMATS.include?(encoding_format)
      raise ArgumentError, "You can only use hexdigest, base64digest or just digest for an encoding format."
    end

    # This lets us dynamically pass in digest_bits as well as encoding formats to the Digest class.
    Digest::SHA2.new(digest_bits).method(encoding_format.to_sym).(string)
  end
  
  def self.openssl_sha2_string(digest_bits: , string: )
    OpenSSL::Digest.digest(digest_bits, string)
  end
end
